collaboration-concept.md

```diff
@@ -0,0 +1,151 @@
+# Collaboration Concept
+
+Collaboration is an often requested for feature in Theia.
+In this document we analyze the collaboration feature of VS Code, specify technical requirements and architecture for Theia and give an overview over the API concept.
+
+## Microsoft VSCode Live Share Extension
+
+The official Microsoft [Live Share Extension](https://visualstudio.microsoft.com/services/live-share/) implements collaboration support for the following features:
+
+- Editor Syncing (Text + Client markers for each user)
+- Host Syncing (i.e. other users follow the host into their currently selected file + position)
+- Terminal Syncing (if enabled). Both input+output is synced
+- Debug Syncing (if enabled). Both users and host can step through the code as if it were on one machine. Breakpoints are synced as well.
+- Plugin Syncing (i.e. external user has access to all installed plugins/extensions of the host user)
+- Chat Functionality
+- Mechanism for making the API available to other extensions, see [vsls](https://www.npmjs.com/package/vsls)
+
+Similar to the tunnel feature, the connection between host and other users can be accomplished using hole-punching.
+I.e. a single central server serves as an intermediary to exchange messages. No messages are being exchanged directly between users.
+If possible - for example, if the users are on the same network - the extension will try to connect directly to the host,
+see [here](https://learn.microsoft.com/en-us/visualstudio/liveshare/reference/connectivity#connection-modes).
```

Comment:
I don't understand the "i.e." following the mention of "hole punching" as this term means establishing a direct connection through each peer's NAT by "punching holes" with the help of an intermediary server to know where to hit. I've also successfully managed to do hole punching in the past using UDP while TCP was more finicky (still worked). The implication means we'd need a UDP transport for direct connections? But that's only if you truly want to do hole punching over the internet.

Reply:
Hole punching might not be the correct term here, right. In the end, all that we need to accomplish is that users can communicate with each other through some means, be it some intermediary server or whatever.

Reply:
Fine if your goal is to abstract the API from how the communication might happen. Although properly laying down the expected modes of communication would help in designing said API: P2P vs Server-centric, how to handle auth/identity/permissions, etc...

If the goal is to be highly generic then the act of creating sessions must be defined at least partially by the embedding layer for the target IDE:
1. Define the communication implementation (a.k.a "driver"? i.e. connect to some server and request a new room)
2. Said driver may implement auth (i.e. prompt the user for credentials in the IDE? defer to some OAuth service?)